File: clients/go/tests/events_test.go
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package scope_test

import (
	"context"
	"sync"
	"testing"
	"time"

	"multiscope/clients/go/clienttesting"
	"multiscope/clients/go/remote"
	"multiscope/clients/go/scope"
	pb "multiscope/protos/tree_go_proto"

	"github.com/google/go-cmp/cmp"
)

func TestEvents(t *testing.T) {
	clt, err := clienttesting.Start()
	if err != nil {
		t.Fatal(err)
	}
	textWriter, err := remote.NewTextWriter(clt, "EventReceiver", nil)
	if err != nil {
		t.Fatal(err)
	}
	received := false
	wg := sync.WaitGroup{}
	wg.Add(1)
	cb := func(ev *scope.Event) error {
		var want []string = textWriter.Path()
		var got = ev.GetPath().GetPath()
		if !cmp.Equal(want, got) {
			t.Errorf("wrong path: got %v but want %v", got, want)
		}
		received = true
		wg.Done()
		return nil
	}
	queue := clt.EventsManager().NewQueueForPath(textWriter.Path(), cb)
	defer queue.Delete()
	// When Register returns, it is not guaranteed the full pipeline between the client
	// and the server is completely setup.
	// Specifically, the client can have a connection but the callback may not be
	// registered to the server.
	time.Sleep(5 * time.Second)
	ctx := context.Background()
	req := &pb.SendEventsRequest{
		Events: []*pb.Event{
			{
				Path: textWriter.Path().NodePath(),
			}},
	}
	_, err = clt.TreeClient().SendEvents(ctx, req)
	if err != nil {
		t.Fatalf("cound not send the event: %v", err)
	}
	wg.Wait()
	if !received {
		t.Error("event has not been received.")
	}
}

File: clients/python/multiscope/remote/events/events_test.py
# Copyright 2023 DeepMind Technologies Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import threading
import time

from absl.testing import absltest

from multiscope.remote import server as multiscope
from multiscope.remote import stream_client
from multiscope.remote.events import events

from multiscope.protos import ticker_pb2
from multiscope.protos import tree_pb2 as pb


def _to_event_pb(path, msg):
  event = pb.Event()
  event.path.path.extend(path)
  event.payload.Pack(msg)
  return event


class EventsTest(absltest.TestCase):

  def setUp(self):
    super().setUp()
    multiscope.start_server(port=0)

  def testSubscribe(self):
    """Tests if we can subscribe to events."""

    wait = threading.Lock()
    wait.acquire()

    collected_action = None

    def collect_event(action):
      nonlocal collected_action
      collected_action = action
      wait.release()

    # Register the callback
    path = ["this", "is", "a", "path"]
    stream_client.GlobalClient().Events().register_ticker_callback(
        path, collect_event)
    # Create an event
    data = ticker_pb2.TickerAction(command=ticker_pb2.Command.CMD_RUN)
    event = _to_event_pb(path, data)
    tree_id = stream_client.GlobalClient().TreeID()
    req = pb.SendEventsRequest(tree_id=tree_id)
    req.events.append(event)

    stream_client.GlobalClient().TreeClient().SendEvents(req)

    wait.acquire()
    self.assertIsNotNone(collected_action)
    self.assertEqual(data, collected_action)


if __name__ == "__main__":
  absltest.main()
